### PR TITLE
Model trial training

### PR DESCRIPTION
**Pull request by Yamini-26**
* Have made a few changes in the LECSFormer model code to bring it to working condition, installing all required packages and matching the versions of all packages. 
* Downloaded and used the 80 images with masks dataset from the NHA12D_A New Pavement Crack Dataset and a Comparison Study Of Crack Detection Algorithms paper.
* Divided the dataset into training and validation sets and created corresponding train.txt and val.txt base filenames.
* Some additional preprocessing steps to match with the dataset included, like - dimensionality reduction and image flipping, and array shape check to match the transpose operation in the np2Tensor.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,5 @@
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Comment (chaozheUB) [inline README.md]: you need to remove this.

File: configs/config.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+MODEL:
+  TYPE: LECSFormer
+  NAME:
+  LECSFormer:
+    EMBED_DIM: 64
+    DEPTHS: [2, 2, 2, 2]
+    NUM_HEADS: [2, 4, 8, 16]
+
```
Comment (chaozheUB) [inline configs/config.yaml]: make line 8 ane empty line.

File: datasets/CrackLS80/images/APFD_01.jpg
Comment (chaozheUB) [inline datasets/CrackLS80/images/APFD_01.jpg:1]: I suggest not adding the figures as part of the repo, but may be put it somewhere say on the box.
https://buffalo.box.com/s/7msenudm0yws4k17j24ym6dsznnd8v7n

**Review by chaozheUB — COMMENTED** [review]
Please add a .gitignore and add `*.pyc` and `__pycache`.

**Comment by Yamini-26** [comment]
I've made all the suggested changes, could you please review it again?

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+__pycache__/
```
Comment (chaozheUB) [inline .gitignore]: please add one empty line to the end of the file.
Also you may need `*__pycache__/` as there are still files like this.

File: datasets/CrackLS80/images/APFD_01.jpg
Comment (chaozheUB) [inline datasets/CrackLS80/images/APFD_01.jpg:1]: As I mentioned previously, you want to move this dataset to the folder.
But please Include in the readme where to find them and how the training data folder structures be for the training code to run.

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+tqdm
+ml-collections
+medpy
+scipy
+h5py
+opencv-python
+einops
+timm
+yacs
+# torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
```
Comment (chaozheUB) [inline requirements.txt]: add an emtpy line to the end.

File: test.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+import torch.nn as nn
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from datasets.crack_datasets import Crack_Datasets
+from networks.LECSFormer import LECSFormer
+from config import get_config
+from collections import OrderedDict
+import cv2
+
+parser = argparse.ArgumentParser()
```
Comment (chaozheUB) [inline test.py:27]: may be put this to a `__main__` function and put to the end?

File: test.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import torch
+import torch.backends.cudnn as cudnn
+import torch.nn as nn
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from datasets.crack_datasets import Crack_Datasets
+from networks.LECSFormer import LECSFormer
+from config import get_config
+from collections import OrderedDict
+import cv2
```
Comment (chaozheUB) [inline test.py]: Please organize the import according to 
first group: default python package (sys, os, etc)
second group: installed package (numpy, pandas, torch)
third group: package in this repo

In each group, the package should be in alphabetical order.
You may ask chat gpt to reorganize this for you.

File: train.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+# from symbol import parameters
+from webbrowser import get
+import numpy as np
+import torch
+import torch.backends.cudnn as cudnn
+from config import get_config
+from networks.LECSFormer import LECSFormer
+from trainer import trainer
+
+def parse_args():
```
Comment (chaozheUB) [inline train.py:24]: same here.

File: train.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                        qkv_bias=config.MODEL.LECSFormer.QKV_BIAS,
+                        qk_scale=config.MODEL.LECSFormer.QK_SCALE,
+                        drop_rate=config.MODEL.DROP_RATE,
+                        drop_path_rate=config.MODEL.DROP_PATH_RATE,
+                        patch_norm=config.MODEL.LECSFormer.PATCH_NORM,
+                        use_checkpoint=config.TRAIN.USE_CHECKPOINT
+                        ).cuda()
+
+
+    trainer(args, model, config)
```
Comment (chaozheUB) [inline train.py]: add an empty line to the end.

**Review by chaozheUB — COMMENTED** [review]
Please also do a [flake8](https://flake8.pycqa.org/en/latest/) check on your program.
You need to install flake8 probably.
You may add a file name `.flake`  with the following to ignore
```
[flake8]
ignore = E261, E402, E501, W293, W504, F405, F403, E101
exclude = .git, .venv, build, cmake-build-*, **/temp/, *results/, requirements.txt
```

**Comment by Yamini-26** [comment]
I've made all suggested changes - Included the flake8, resolved a huge bunch of existing flake8 errors, moved the dataset folder to the Box, updated the structure and usage instructions of the dataset in the README file, and sorted all the imports. 
Please let me know if it is all okay to proceed.

**Review by chaozheUB — COMMENTED** [review]
Looks good! 
Are these codes able to run? 
If so I suggest you squash all the commits to one and we can merge this version

File: try.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
+# import os
+# import cv2
```
Comment (chaozheUB) [inline try.py]: what does this script do?

File: requirements.txt (hunk 4 above)
Comment (chaozheUB) [reply requirements.txt]: ^

File: output/crackls80/log.txt [hunk 10]
```diff
@@ -0,0 +1,32 @@
+[22:01:31.646] Namespace(root_path='datasets/CrackLS315', num_classes=1, output_dirs='output/crackls315', max_epochs=100, batch_size=16, n_gpu=2, deterministic=1, base_lr=0.001, img_size=[512, 512], seed=44, cfg='configs/config.yaml', resume=None, accumulation_steps=None, cache_mode='part', opts=None, zip=False, use_checkpoint=False, amp_opt_level='O1', tag=None, eval=False, throughput=False)
```
Comment (chaozheUB) [inline output/crackls80/log.txt:1]: why do we need this?

File: networks/__pycache__/LECSFormer.cpython-311.pyc
Comment (chaozheUB) [inline networks/__pycache__/LECSFormer.cpython-311.pyc:1]: why this is still here?

File: try.py (hunk 9 above)
Comment (Yamini-26) [reply try.py]: This is just to try things separately. I can remove this from the repo as well.

File: output/crackls80/log.txt (hunk 10 above)
Comment (Yamini-26) [reply output/crackls80/log.txt:1]: It's like a checkpoint to see which epoch exactly did the code stop running, or from where the model started getting overfit on the data, and so on.

File: networks/__pycache__/LECSFormer.cpython-311.pyc
Comment (Yamini-26) [reply networks/__pycache__/LECSFormer.cpython-311.pyc:1]: It's just the cached version, will remove that as well.

**Comment by Yamini-26** [comment]
Yes Professor, this version of code is able to run.

File: try.py (hunk 9 above)
Comment (chaozheUB) [reply try.py]: OK please remove unnecessary ones.

File: output/crackls80/log.txt (hunk 10 above)
Comment (chaozheUB) [reply output/crackls80/log.txt:1]: OK, I assume there will be more like this in the future? Then you may want to have a log name with some date info,

**Comment by chaozheUB** [comment]
> Yes Professor, this version of code is able to run.

OK, then pleas squash to 2 commits and we can merge it.
If possible, you should have the first commit from the LECSFormer as it is, then the second commit contains your update. 
It may need some rework but for logging purpose it would worth it. After all we want the commit in master to carry clear meanings.

File: output/crackls80/log.txt (hunk 10 above)
Comment (Yamini-26) [reply output/crackls80/log.txt:1]: Yes, that's right. Since the training is not able to run on my system, due to insufficient GPU memory there is only one for now.

**Comment by Yamini-26** [comment]
> > Yes Professor, this version of code is able to run.
> 
> OK, then pleas squash to 2 commits and we can merge it. If possible, you should have the first commit from the LECSFormer as it is, then the second commit contains your update. It may need some rework but for logging purpose it would worth it. After all we want the commit in master to carry clear meanings.

Sure, I'll try and keep the LECSFormer as the first commit and add our changes as the second one.

**Comment by Yamini-26** [comment]
Hi professor, since some files and folder like try.py, and the dataset folder are not ignored in the previous version, but now that I have added all of those in the .gitignore, it is causing a conflict while trying to squash the commits. 
I keep getting this error - error: The following untracked working tree files would be overwritten by merge:
        try.py
Please move or remove them before you merge.
Aborting
hint: Could not execute the todo command
hint:
hint:     squash 570431836e8440a00f9f2674d9a4b9204f922b2f LECSFormer
hint:
hint: It has been rescheduled; To edit the command before continuing, please
hint: edit the todo list first:
hint:
hint:     git rebase --edit-todo
hint:     git rebase --continue

I tried 'git rm' to remove the cached versions, I pushed them all to 'stash', even though all those are clear, it still keeping throwing the same error.
So, I ran 'git clean -fd' to remove the files/directory that are untracked. This command removes all these files and folders in my working directory. Only then I am able to continue with the squashing. I would have to replace all these files and folders manually after squashing, and I also am not sure if this will occur next time as well while pushing or squashing. 
Can you please help me figure out what can be a permanent solution to this problem?

**Comment by chaozheUB** [comment]
I am not familar with this error. You may have to look it up. Normally for such kind of activities, I would recommend the following.

1.  start an whole new branch from the current master, which should be empty.
2. manually added the raw code from the original repo that you start from, you should also add the .gitignore, .flake8 and so on. Make a first commit to this new branch
3. rebase your current branch `model-trial-training` on to this new branch, clean any conflicts.
4. squash all the commits that is on top of the new branch commit to 1 commit. 
 
This is only one way to do this.

File: .flake8 [hunk 11]
```diff
@@ -0,0 +1,3 @@
+[flake8]
+ignore = E261, E402, E501, W293, W504, F405, F403, E101, F841
+exclude = .git, .venv, build, cmake-build-*, **/temp/, *results/, requirements.txt
```
Comment (chaozheUB) [inline .flake8:3]: Please update this to your virtual environment name (`.venv` was the name for my case), and the corresponding folders you want to ignore the flake8.

File: configs/config.yaml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+MODEL:
+  TYPE: LECSFormer
+  NAME:
+  LECSFormer:
+    EMBED_DIM: 64
+    DEPTHS: [2, 2, 2, 2]
+    NUM_HEADS: [2, 4, 8, 16]
```
Comment (chaozheUB) [inline configs/config.yaml:7]: Add a new line

File: test.py (hunk 5 above)
Comment (chaozheUB) [reply test.py:27]: Please add the parser to `__main__` function
